v1alpha2: Implementation

gaocegege opened the pull request:
@ScorpioCPH and @gaocegege implemented the main logic and basic unit test cases (73% coverage).

I think it is time to open the PR, and we decided to add basic e2e test cases for v1alpha2 these days.

We have created a branch for v1alpha1 and when the v1alpha2 is merged into master we will remove the branch v1alpha2.

PS: Suggest to rebase the PR to master, not squash it since the commit message may be helpful for new contributors.

WDYT? @jlewi

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/tf-operator/492)
<!-- Reviewable:end -->

jlewi commented:
Fantastic job on test the test coverage.

jlewi commented:
/assign jlewi

jlewi commented:
This PR is massive can we split it up? Can we have a PR to define the v1alpha2 API and all the generated code?

Following up our conversation in slack; is there anyway we can keep the existing code so that we can build two different binaries 1 for v1alpha1 and 1 for v1alpha2?

Our test/release infrastructure isn't really set up to deal with branches. My concern is that it will take us some time to update all the downstream code (e.g. examples, kubeflow/kubeflow) . It would be nice if in that time we could continue to build and test controllers for both versions so that we could do the updates incrementally.

Is there a reason we can't just have two sets of packages? e.g.
pkg/controller
pkg/controllerv2
pkg/trainer
pkg/trainerv2
cmd/tfoperator/app
cmd/tfoperator/appv2

gaocegege commented:
@jlewi 

I kept the v1alpha1 code and added v1alpha2 code as v2. I think it will help us update the downstream code, and PTAL

If you think the code LGTY, I suggest rebasing it into master to keep a clear git commit history.

ScorpioCPH commented:
@gaocegege Thanks! Checkout a new branch `v1alpha1` from `master` is a good idea.

gaocegege commented:
I have no idea why the CI failed although argo passes all tests. http://testing-argo.kubeflow.org/workflows/kubeflow-test-infra/kubeflow-tf-operator-presubmit-tfjob-e2e-492-caf21fa-224-a68c?tab=workflow

jlewi commented:
@gaocegege Thanks the v2 stuff looks good to me. If you want to rebase on master to preserve commit history that seems good to me. 

The reason the argo workflow passes even though there are test failures is because we don't necessarily turn test failures into step failures. For example, if there are lint/unit test failures those will be reported as junit errors, but the step in argo will still be reported as successful because the python script didn't exit with an error.

gaocegege commented:
@jlewi 

Thanks for your explanation. Are the errors imported by this PR or the previous PR? If they are I can fix it. If not I think we could merge it ASAP since there are some PRs need to be rebased the the PR.

gaocegege commented:
The simple-tfjob and gpu test time out, and I can not see the log because of #496 

Could we rebase it into master and fix the problems soon?

jlewi commented:
@gaocegege @ScorpioCPH See [comment](https://github.com/kubeflow/tf-operator/issues/496#issuecomment-375496676) for work around to access the logs.

gaocegege commented:
https://console.cloud.google.com/logs/viewer?project=kubeflow-ci&resource=k8s_cluster%2Flocation%2Fus-east1-d%2Fcluster_name%2Fkubeflow-testing&angularJsUrl=%2Flogs%2Fviewer%3Fproject%3Dkubeflow-ci%26resource%3Dk8s_cluster%252Flocation%252Fus-east1-d%252Fcluster_name%252Fkubeflow-testing&authuser=1&minLogLevel=0&expandAll=false&timestamp=2018-03-23T01%3A25%3A52.731000000Z&interval=NO_LIMIT&filters=text%3Akubeflow-tf-operator-presubmit-tfjob-e2e-492-b317aef-225-99c4&scrollTimestamp=2018-03-22T12%3A24%3A44.230913000Z&advancedFilter=resource.type%3D%22k8s_cluster%22%0Aresource.labels.project_id%3D%22kubeflow-ci%22%0Aresource.labels.cluster_name%3D%22kubeflow-testing%22%0Aresource.labels.location%3D%22us-east1-d%22%0Atimestamp%3D%222018-03-22T12%3A24%3A44.131712000Z%22%0AinsertId%3D%221amkekse7ey87%22

"workflows.argoproj.io "kubeflow-tf-operator-presubmit-tfjob-e2e-492-b317aef-225-99c4" not found"

gaocegege commented:
/retest

gaocegege commented:
/retest

gaocegege commented:
/retest

gaocegege commented:
/retest

gaocegege commented:
/retest

gaocegege commented:
Insufficient regional quota to satisfy request for resource: "DISKS_TOTAL_GB". The request requires '100.0' and is short '4.0'. The regional quota is '4096.0' with '96.0' available.

I will have a try when the cluster is free.

gaocegege commented:
/retest

gaocegege commented:
/retest

gaocegege commented:
@jlewi Could we merge it ASAP since it blocks the continuous development of v1alpha2?

jlewi commented:
The disk quota issue is kubeflow/testing#81 and should be fixed.

jlewi commented:
Can we split the PR please and commit the code in v1alpha2 and code gen first?

Massive PRs slow down code review and block other work. Splitting a PR into multiple PRs is a great way to speed up the review process.

jlewi commented:
Review status: 0 of 49 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[cmd/tf-operator.v2/app/server.go, line 53 at r1](https://reviewable.io:443/reviews/kubeflow/tf-operator/492#-L8K_aie3CgbyJINfwYd:-L8K_aie3CgbyJINfwYe:bh4agvd) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/cmd/tf-operator.v2/app/server.go#L53)):*
> ```Go
> const RecommendedKubeConfigPathEnv = "KUBECONFIG"
> 
> func Run(opt *options.ServerOption) error {
> ```

This looks very similar to 
https://github.com/kubeflow/tf-operator/blob/master/cmd/tf-operator/app/server.go

Only difference I see is we get rid of reading readControllerConfig and more importantly we use the v1alpha packages

Why not refactor the code a bit to promote code reuse?

What if we had a  RunForVersionFunction

that could be called by Run.

I think this would be useful because this won't be the only time we change the API. For example at a minimum at sometime in the future we will have a v1beta1 and a v1 API.

---

*[cmd/tf-operator.v2/app/options/options.go, line 15 at r1](https://reviewable.io:443/reviews/kubeflow/tf-operator/492#-L8Kad_r2_4lDmPxDnSB:-L8Kad_r2_4lDmPxDnSC:b-y9bb6m) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/cmd/tf-operator.v2/app/options/options.go#L15)):*
> ```Go
> // limitations under the License.
> 
> package options
> ```

Could we just reuse the existing options package?
Is there anything different about this file?

---

*[examples/simple_tf_job.yaml, line 1 at r1](https://reviewable.io:443/reviews/kubeflow/tf-operator/492#-L8Kaqtd-FbAGw4UcUH3:-L8Kaqtd-FbAGw4UcUH4:bghxr81) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/examples/simple_tf_job.yaml#L1)):*
> ```YAML
> apiVersion: "kubeflow.org/v1alpha2"
> ```

Why do we need this?

---

*[examples/tf_job_v1alpha2.yaml, line 1 at r1](https://reviewable.io:443/reviews/kubeflow/tf-operator/492#-L8KaxZV8P8zGjvFb65e:-L8KaxZV8P8zGjvFb65f:byquu61) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/examples/tf_job_v1alpha2.yaml#L1)):*
> ```YAML
> apiVersion: "kubeflow.org/v1alpha2"
> ```

Do we need this?

---

*[pkg/controller.v2/controller.go, line 142 at r1](https://reviewable.io:443/reviews/kubeflow/tf-operator/492#-L8KdiA3-pJnKimEcXOE:-L8KdiA3-pJnKimEcXOF:b349lox) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/pkg/controller.v2/controller.go#L142)):*
> ```Go
> 	// - "tf-operator/tfjob-abc/worker/services", expects 4 adds.
> 	// - "tf-operator/tfjob-abc/worker/pods", expects 4 adds.
> 	expectations ControllerExpectationsInterface
> ```

Can you explain this design?  Why do we need to check number of events? Why don't we just check that if we expect workers 1....N that there are 1...N pods as expected?

Are there other K8s controllers that are designed this way?

---

*[pkg/controller.v2/controller_pod.go, line 67 at r1](https://reviewable.io:443/reviews/kubeflow/tf-operator/492#-L8KjJrmpPC07kSL9ZJe:-L8KjJrmpPC07kSL9ZJf:b-z7rv2a) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/pkg/controller.v2/controller_pod.go#L67)):*
> ```Go
> 
> 		expectationPodsKey := genExpectationPodsKey(tfjobKey, rt)
> 		err := tc.expectations.ExpectCreations(expectationPodsKey, diff)
> ```

Why are you only checking the number of creations? Why don't you check that the actual pods you expect exist?

Suppose we expect 4 worker replicas; 1...N.

But suppose we get the following create events
Worker 0
Worker 0
Worker 1
Worker 2


So worker 0 is created twice but worker 3 is never created.

Won't the code think all 4 workers have been created even though its not true?

---

*[pkg/controller.v2/controller_utils.go, line 398 at r1](https://reviewable.io:443/reviews/kubeflow/tf-operator/492#-L8Kira-hohu9Q3pvUsA:-L8Kira-hohu9Q3pvUsB:b-kaa3nv) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/pkg/controller.v2/controller_utils.go#L398)):*
> ```Go
> 	// CreatePodsOnNode creates a new pod according to the spec on the specified node,
> 	// and sets the ControllerRef.
> 	CreatePodsOnNode(nodeName, namespace string, template *v1.PodTemplateSpec, object runtime.Object, controllerRef *metav1.OwnerReference) error
> ```

Why do we need a function to create pods on specific nodes?

---

*[pkg/util/signals/signal.go, line 2 at r1](https://reviewable.io:443/reviews/kubeflow/tf-operator/492#-L8KkEt43DCroFkSnA-Z:-L8KkEt43DCroFkSnA-_:b-ydllmw) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/pkg/util/signals/signal.go#L2)):*
> ```Go
> /*
> Copyright 2017 The Kubernetes Authors.
> ```

Can you move this into a separate PR? It will speed up the review process if we can break this up into small PRs so its easier to focus on the important bits in each PR.

---

*[pkg/util/signals/signal_windows.go, line 17 at r1](https://reviewable.io:443/reviews/kubeflow/tf-operator/492#-L8KjuizsWBKQXGeF0wB:-L8KjuizsWBKQXGeF0wC:b-8f55tg) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/pkg/util/signals/signal_windows.go#L17)):*
> ```Go
> */
> 
> package signals
> ```

Why do we have a package for windows? When would we ever run this code on windows?

---

*[test/e2e/sleep-and-random-exit/Dockerfile, line 3 at r1](https://reviewable.io:443/reviews/kubeflow/tf-operator/492#-L8Kk4wGDPKjXz30QnpF:-L8Kk4wGDPKjXz30QnpG:bf92tn7) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/test/e2e/sleep-and-random-exit/Dockerfile#L3)):*
> ```Dockerfile
> FROM ubuntu:16.04
> 
> COPY ./sleep_and_random_exit.sh ~/
> ```

If we need this, please move this into a separate PR. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubeflow/tf-operator/492)*
<!-- Sent from Reviewable.io -->

gaocegege commented:
@jlewi API and code-gen is https://github.com/kubeflow/tf-operator/pull/492/commits/0d61bf91038f3d8dcfec5917d4c425364844cb7d

V1alpha2 core logic is https://github.com/kubeflow/tf-operator/pull/492/commits/4e1dd8d297baadd2a4a66af252b6112fa510635a

These two commits are the most important commits and other commits mainly focus on test and I think we could just look through them. 

Sorry that I think it is hard to split the PRs and here is the reason:

We removed the v1alpha1 during the development and added it again in one commit since you recommend to keep two implementations in master. Then if we split the PR, there are lots of work to do to keep two implementations from the beginning instead of the commit https://github.com/kubeflow/tf-operator/pull/492/commits/caf21fad8937032a3301a1feb9b22f9d58ef3445.

Or we could keep one implementation and just remove v1alpha1 then I think I could split the previous commits since the commits in v1alpha2 already do.

Another option is to keep the development in v1alpha2 and wait until the downstream is ready, then we could split the commits in v1alpha2 and overwrite master.

/cc @ScorpioCPH

jlewi commented:
I really think this PR needs to be split up. There's a bunch of code in here that should be relatively straightforward to approve ; e.g. signals. So the inclusion in this PR is just a distraction.

I don't understand why splitting up this PR is difficult.

Create a new based of master head.
Copy the files 
pkg/apis/tensorflow/v1alpha2 
pkg/clientset/.../v1alpha2 
into the directory
Create a PR.

That PR would be just the files for the API which should be relatively straightforward to review.

Follow the same process to create a PR for the signals package.

Once those PRs are committed which should be relatively straightforward. You can sync with master. At that point the only thing left is the controller package which is where I think the most attention needs to be paid.

ScorpioCPH commented:
@Jimexist You can check out branch v1alpha2, the whole commit history are there.

jlewi commented:
@ScorpioCPH I'm not sure I follow. The point of splitting up this PR is that the vast majority of it is straightforward to review. So if we split up the PR we can get most of the files committed pretty quickly.

This would leave us with the controller logic and the controller logic is where I'd like to focus during the review process because there appear to be substantial changes to how we track and reconcile resources as part of the TFJob.

jlewi commented:
Review status: 0 of 49 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/controller.v2/controller.go, line 76 at r1](https://beta.reviewable.io/reviews/kubeflow/tf-operator/492#-L8_2TtJBgF5hn96poWD:-L8_2TtJBgF5hn96poWE:bcn9n9t) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/pkg/controller.v2/controller.go#L76)):*
> ```Go
> 	// wait between two reconciler sync.
> 	// It is set to 15 sec by default.
> 	// TODO(cph): maybe we can let it grows by multiple in the future
> ```

Do we really need exponential backoff?

If we correctly configure and listen to all appropriate events for a TFJob then we should be able to set the periodic reconcile to a large value 1-2 minutes.

---

*[pkg/controller.v2/controller.go, line 197 at r1](https://beta.reviewable.io/reviews/kubeflow/tf-operator/492#-L8_3GV-8N-fpv0hXx0x:-L8_3GV-8N-fpv0hXx0y:b-ak36oo) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/pkg/controller.v2/controller.go#L197)):*
> ```Go
> 
> 	// Create tfjob informer.
> 	tfJobInformer := tfJobInformerFactory.Kubeflow().V1alpha2().TFJobs()
> ```

Do we configure this informer to periodically generate update events to be used for reconcilation like we currently do?
If we do where does that happen?
If not why not?
https://github.com/kubeflow/tf-operator/blob/master/cmd/tf-operator/app/server.go#L82

---

*[pkg/controller.v2/controller.go, line 289 at r1](https://beta.reviewable.io/reviews/kubeflow/tf-operator/492#-L8_4o724HsBeTRfO56E:-L8_4o724HsBeTRfO56F:busk3pq) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/pkg/controller.v2/controller.go#L289)):*
> ```Go
> // attempt to process it, by calling the syncHandler.
> func (tc *TFJobController) processNextWorkItem() bool {
> 	key, quit := tc.workQueue.Get()
> ```

Can I suggest making processNextWorkItem a utility function that can be called from the v1 and v2 controller?

I just fixed some issues in #501. I don't think this code is actually different so it would be useful to avoid code duplication.

---

*[pkg/controller.v2/controller.go, line 325 at r1](https://beta.reviewable.io/reviews/kubeflow/tf-operator/492#-L8_5aobDtjqSHVvaiyM:-L8_5aobDtjqSHVvaiyN:b-o5unzg) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/pkg/controller.v2/controller.go#L325)):*
> ```Go
> 	startTime := time.Now()
> 	defer func() {
> 		log.Infof("Finished syncing tfjob %q (%v)", key, time.Since(startTime))
> ```

It would be good to use logrus and add fields to tag log entries with job name. 
Like
https://github.com/kubeflow/tf-operator/blob/master/pkg/controller/controller.go#L221

This will make it alot easier to debug issues processing jobs.

---

*[pkg/controller.v2/controller.go, line 344 at r1](https://beta.reviewable.io/reviews/kubeflow/tf-operator/492#-L8_5MWX5b3dFew73ZIX:-L8_5MWX5b3dFew73ZIY:bqh3tps) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/pkg/controller.v2/controller.go#L344)):*
> ```Go
> 
> 	tfjob := sharedtfjob.DeepCopy()
> 	tfjobNeedsSync := tc.satisfiedExpectations(tfjob)
> ```

What is satisifiedExpections? Why don't we just call tc.reconcileTFJobs and let the reconcile function figure out what if anything needs to be done?

---

*[pkg/controller.v2/controller.go, line 362 at r1](https://beta.reviewable.io/reviews/kubeflow/tf-operator/492#-L8_5zPo8CCeVX_cyczO:-L8_5zPo8CCeVX_cyczP:bpkz1k0) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/pkg/controller.v2/controller.go#L362)):*
> ```Go
> func (tc *TFJobController) reconcileTFJobs(tfjob *tfv1alpha2.TFJob) error {
> 
> 	pods, err := tc.getPodsForTFJob(tfjob)
> ```

Why do we need to get the pods and service here?
Can't we just rely on reconcilePods and reconcileServices to do the correct thing?

---

*[pkg/controller.v2/controller_pod.go, line 37 at r1](https://beta.reviewable.io/reviews/kubeflow/tf-operator/492#-L8_7FMD0FRUPEHh58rw:-L8_7FMD0FRUPEHh58rx:b-6mcpfm) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/pkg/controller.v2/controller_pod.go#L37)):*
> ```Go
> // It will requeue the tfjob in case of an error while creating/deleting pods.
> func (tc *TFJobController) reconcilePods(
> 	tfjob *tfv1alpha2.TFJob,
> ```

When a job is finished but not deleted, what prevents pods from being recreated?

---

*[pkg/controller.v2/controller_pod.go, line 242 at r1](https://beta.reviewable.io/reviews/kubeflow/tf-operator/492#-L8_6xXj9KrHj9KN4po-:-L8_6xXj9KrHj9KN4po0:b-vgpd6q) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/pkg/controller.v2/controller_pod.go#L242)):*
> ```Go
> }
> 
> // When a pod is created, enqueue the tfjob that manages it and update its expectations.
> ```

Why do we need to update the expectations? What are the expectations used for?

---

*[pkg/controller.v2/controller_pod.go, line 293 at r1](https://beta.reviewable.io/reviews/kubeflow/tf-operator/492#-L8_4A4cE7lr8epSAW2Z:-L8_4A4cE7lr8epSAW2_:b-bvdsn6) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/pkg/controller.v2/controller_pod.go#L293)):*
> ```Go
> // and new replica set. old and cur must be *v1.Pod types.
> func (tc *TFJobController) updatePod(old, cur interface{}) {
> 	// TODO(CPH): handle this gracefully.
> ```

Can we just enqueue a work item for the associated job which will cause us to call reconcile for the job?

We can rely on the workQueue to aggregate events and rate limit things.

---

*[pkg/controller.v2/controller_service.go, line 172 at r1](https://beta.reviewable.io/reviews/kubeflow/tf-operator/492#-L8_6Vlc9ZR6yZjSRsFD:-L8_6Vlc9ZR6yZjSRsFE:bwnrazr) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/pkg/controller.v2/controller_service.go#L172)):*
> ```Go
> 	}
> 
> 	// If any adoptions are attempted, we should first recheck for deletion
> ```

What are adoptions?

---

*[pkg/controller.v2/controller_service.go, line 255 at r1](https://beta.reviewable.io/reviews/kubeflow/tf-operator/492#-L8_4ZQD5rYVXSHVlivc:-L8_4ZQD5rYVXSHVlivd:b-bvdsn6) ([raw file](https://github.com/kubeflow/tf-operator/blob/0947506abf7d81ab2639a2d590973dc426cc096e/pkg/controller.v2/controller_service.go#L255)):*
> ```Go
> // and new replica set. old and cur must be *v1.Service types.
> func (tc *TFJobController) updateService(old, cur interface{}) {
> 	// TODO(CPH): handle this gracefully.
> ```

Can we just enqueue a work item for the associated job which will cause us to call reconcile for the job?

We can rely on the workQueue to aggregate events and rate limit things.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/kubeflow/tf-operator/492)*
<!-- Sent from Reviewable.io -->

jlewi commented:
@gaocegege How's it going with v1alpha2?

gaocegege commented:
@jlewi 

We planned to discuss about the integration but that community meeting has no time for this topic. Maybe we could schedule a meeting for it

/cc @DjangoPeng @ScorpioCPH @ddysher

DjangoPeng commented:
SGTM.

We could schedule a online meeting to discuss and make a decision after tomb-sweeping day (the traditional Qingming festival). How about Apr 8th?

@gaocegege @ddysher @ScorpioCPH WDYT

ddysher commented:
For anyone who happens to read this PR, we will split the PR into smaller ones to ease review.

- https://github.com/kubeflow/tf-operator/pull/523
 -https://github.com/kubeflow/tf-operator/pull/526
